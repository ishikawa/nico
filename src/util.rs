use std::cell::RefCell;
use std::rc::Rc;

pub fn wrap<T>(ty: T) -> Rc<RefCell<T>> {
    Rc::new(RefCell::new(ty))
}

pub trait UniqueNaming {
    fn next(&mut self) -> String;
}

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug)]
pub struct SequenceNaming {
    type_var_index: usize,
    prefix: String,
}

impl SequenceNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }
}

impl UniqueNaming for SequenceNaming {
    fn next(&mut self) -> String {
        let i = self.type_var_index;
        self.type_var_index += 1;
        format!("{}{}", self.prefix, i)
    }
}
