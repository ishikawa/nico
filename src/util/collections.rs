use crate::arena::BumpaloVec;
use std::cell::Ref;

#[derive(Debug)]
pub struct RefVecIter<'r, 'arena, T> {
    vec: Ref<'r, BumpaloVec<'arena, T>>,
}

impl<'r, 'arena, T> RefVecIter<'r, 'arena, T> {
    pub fn new(vec: Ref<'r, BumpaloVec<'arena, T>>) -> Self {
        Self { vec }
    }
}

impl<'a: 'r, 'r, 'arena, T> IntoIterator for &'a RefVecIter<'r, 'arena, T> {
    type IntoIter = std::slice::Iter<'r, T>;
    type Item = &'r T;

    fn into_iter(self) -> Self::IntoIter {
        self.vec.iter()
    }
}
