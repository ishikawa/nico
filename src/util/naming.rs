use std::collections::HashMap;

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug, Default)]
pub struct SequenceNaming {
    names: HashMap<String, i32>,
}

impl SequenceNaming {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn next<S: AsRef<str>>(&mut self, name: S) -> String {
        let name = name.as_ref();
        let n = self.names.entry(name.to_string()).or_insert(0);
        let next = format!("{}.{}", name, n);

        *n += 1;
        next
    }
}

/// This struct produces a unique name while giving a sequential
/// number to the name passed to the next() method.
#[derive(Debug)]
pub struct PrefixNaming {
    type_var_index: usize,
    prefix: String,
}

impl PrefixNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }

    pub fn next(&mut self) -> String {
        let i = self.type_var_index;
        let next = format!("{}{}", self.prefix, i);

        self.type_var_index += 1;
        next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prefix_naming() {
        let mut naming = PrefixNaming::new("my_");

        assert_eq!(naming.next(), "my_0".to_string());
        assert_eq!(naming.next(), "my_1".to_string());
    }

    #[test]
    fn sequence_naming() {
        let mut naming = SequenceNaming::new();

        assert_eq!(naming.next("x"), "x.0".to_string());
        assert_eq!(naming.next("x"), "x.1".to_string());
        assert_eq!(naming.next("y"), "y.0".to_string());
    }
}
