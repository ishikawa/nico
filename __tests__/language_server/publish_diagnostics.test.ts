import { LanguageServer, LanguageServerAgent, spawn_server } from "../util/lsp";
import { filterTestCases, TestCaseBase, readTestFileSync, getTestName } from "../util/testcase";

let server: LanguageServer | undefined;

beforeAll(async () => {
  server = await spawn_server();
});

afterAll(() => {
  if (server) {
    return server.stop();
  }
});

let cases: TestCaseBase[] = [
  // Undefined variable
  {
    input: "a"
  },
  // Defined variable
  {
    // prettier-ignore
    input: [
      "let a = 1",
      "a"
    ].join("\n")
  },
  // Block scope
  {
    // prettier-ignore
    input: [
      "if 3 > 0",
      "    let a = 1",
      "    println_i32(a)",
      "end",
      "a" // `a` should not be visible in this scope.
    ].join("\n")
  },
  {
    // prettier-ignore
    input: [
      "case 3",
      "when x if x == 3",
      "    println_i32(x)",
      "else",
      "    println_i32(10)",
      "end",
      "x" // `x` should not be visible in this scope.
    ].join("\n")
  },
  {
    // prettier-ignore
    input: [
      "let Rectangle = 1",
      "let rect = Rectangle {}", // expected struct
    ].join("\n")
  },
  {
    // prettier-ignore
    input: [
      "fun foo() ->", // missing return type
      "end",
    ].join("\n")
  },
  {
    // prettier-ignore
    input: [
      "fun foo(bar: Int[]) -> Int[]",
      "    let x = bar[0]",
      "    x * 2",  // mismatched types
      "end",
    ].join("\n")
  }
];

filterTestCases(cases).forEach((testCase, i) => {
  let src = readTestFileSync(testCase);
  let name = getTestName(testCase);

  test(`${i}: publishDiagnostics at \`${name}\``, async done => {
    const agent = new LanguageServerAgent(server!, { sequence: i });

    const diagnostics = await agent.openDocument(name, src);
    expect(diagnostics).toMatchSnapshot();

    done();
  });
});
