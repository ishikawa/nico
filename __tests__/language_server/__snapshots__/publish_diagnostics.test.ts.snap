// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0: publishDiagnostics at \`a\` 1`] = `
Array [
  Object {
    "message": "Cannot find name 'a'.",
    "range": Object {
      "end": Object {
        "character": 1,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "nico-ls",
  },
]
`;

exports[`1: publishDiagnostics at \`let a = 1
a\` 1`] = `Array []`;

exports[`2: publishDiagnostics at \`if 3 > 0
    let a = 1
    println_i32(a)
end
a\` 1`] = `
Array [
  Object {
    "message": "Cannot find name 'a'.",
    "range": Object {
      "end": Object {
        "character": 1,
        "line": 4,
      },
      "start": Object {
        "character": 0,
        "line": 4,
      },
    },
    "severity": 1,
    "source": "nico-ls",
  },
]
`;

exports[`3: publishDiagnostics at \`case 3
when x if x == 3
    println_i32(x)
else
    println_i32(10)
end
x\` 1`] = `
Array [
  Object {
    "message": "Cannot find name 'x'.",
    "range": Object {
      "end": Object {
        "character": 1,
        "line": 6,
      },
      "start": Object {
        "character": 0,
        "line": 6,
      },
    },
    "severity": 1,
    "source": "nico-ls",
  },
]
`;

exports[`4: publishDiagnostics at \`let Rectangle = 1
let rect = Rectangle {}\` 1`] = `
Array [
  Object {
    "message": "Expected struct, found 'local variable \`Rectangle\`'.",
    "range": Object {
      "end": Object {
        "character": 20,
        "line": 1,
      },
      "start": Object {
        "character": 11,
        "line": 1,
      },
    },
    "severity": 1,
    "source": "nico-ls",
  },
]
`;

exports[`5: publishDiagnostics at \`fun foo() ->
end\` 1`] = `
Array [
  Object {
    "message": "Syntax Error: expected type annotation",
    "range": Object {
      "end": Object {
        "character": 12,
        "line": 0,
      },
      "start": Object {
        "character": 10,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "nico-ls",
  },
]
`;

exports[`6: publishDiagnostics at \`fun foo(bar: Int[]) -> Int[]
    let x = bar[0]
    x * 2
end\` 1`] = `
Array [
  Object {
    "message": "mismatched types
expected \`Int\`
   found \`Int[]\`",
    "range": Object {
      "end": Object {
        "character": 9,
        "line": 2,
      },
      "start": Object {
        "character": 4,
        "line": 2,
      },
    },
    "severity": 1,
    "source": "nico-ls",
  },
]
`;
